src/adapters/philips_hue/api/mod.rs

```diff
@@ -0,0 +1,230 @@
+    pub id: String,
+    pub ip: String,
+    pub token: String,
+}
+
+impl std::fmt::Display for HueHubApi {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Hue Bridge id:{} at {:?}", self.id, self.ip)
+    }
+}
+
+impl HueHubApi {
+    pub fn new(id: String, ip: String) -> HueHubApi {
+        // TODO: Generate a unique but reproducible access token.
+        // The token must not be guessable by an outsider.
+
+        // TODO: replace with a loooong stored random string
+        let stored_random = String::from("4"); // chosen by fair dice roll.
+        let token_gen = HueHubApiToken { iv: stored_random, id: id.clone() };
+        let token = format!("foxbox-{}", token_gen.as_hash());
+        HueHubApi { id: id, ip: ip, token: token }
+    }
+
+    pub fn get(&self, cmd: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("GET request to Philips Hue bridge {}: {}", self.id, url);
+        let content = http::get(url);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn post(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("POST request to Philips Hue bridge {}: {} data: {}", self.id, &url, &data);
+        let content = http::post(url, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn post_unauth(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/{}", self.ip, cmd);
+        debug!("POST request to Philips Hue bridge {}: {} data: {}", self.id, &url, &data);
+        let content = http::post(url, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn put(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let url = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("PUT request to Philips Hue bridge {}: {} data: {}", self.id, &url, &data);
+        let content = http::put(url, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn is_available(&self) -> bool {
+        let url = format!("http://{}/", self.ip);
+        let content = http::get(url);
+        match content {
+            Ok(value) => {
+                value.contains("hue personal wireless lighting")
+            },
+            Err(_) => {
+                false
+            }
+        }
+    }
+
+    pub fn get_settings(&self) -> String {
+        // [{"error":{"type":1,"address":"/","description":"unauthorized user"}}]
+        self.get("".to_owned()).unwrap_or("".to_owned()) // TODO no unwrap
+    }
+
+    pub fn is_paired(&self) -> bool {
+        let settings = self.get_settings();
+        !settings.contains("unauthorized user")
+    }
```

Comment:
I only half agree. In the current scheme of things it makes sense to move it out, but over its lifetime, the *HueHubApi* object has turned from something that manages its own state for buffering purposes into a thin static layer. It had to because cross-referencing objects turns ugly fast when multi-threading got involved, and given the deadline this week there was not enough time to make it proper.

In the not too long run I want the api and light objects to manage their state again. This will not only enable intelligent buffering again that provides considerable performance gains, it will also pave the way for elegant management of devices popping in and out of existence during runtime.